src/mlpack/methods/neat/neat_impl.hpp

```diff
@@ -0,0 +1,540 @@
+    }
+  }
+
+  return genomeList[maxIdx];
+}
+
+// Speciates the genomes.
+template <class TaskType,
+          class ActivationFunction,
+          class SelectionPolicy>
+void NEAT<TaskType, ActivationFunction, SelectionPolicy>::Speciate(bool init)
+{
+  // Translate the genome into points in space.
+  arma::mat data(Genome<ActivationFunction>::nextInnovID, popSize,
+      arma::fill::zeros);
+  for (size_t i = 0; i < popSize; i++)
+  {
+    for (size_t j = 0; j < genomeList[i].connectionGeneList.size(); j++)
+    {
+      size_t innovID = genomeList[i].connectionGeneList[j].InnovationID();
+      data(innovID, i) = genomeList[i].connectionGeneList[j].Weight();
+    }
+  }
+
+  arma::Row<size_t> assignments(popSize, arma::fill::zeros);
+  kmeans::KMeans<metric::EuclideanDistance, kmeans::SampleInitialization,
+      kmeans::MaxVarianceNewCluster, kmeans::CoverTreeDualTreeKMeans> k; 
+  if (init)
+  {
+    centroids = arma::mat(data.n_rows, numSpecies, arma::fill::zeros);
+    k.Cluster(data, numSpecies, assignments, centroids);
+  }
+  else
+  {
+    centroids.resize(data.n_rows, numSpecies);
+    k.Cluster(data, numSpecies, assignments, centroids, false, true);
+  }
+
+  // Clear the old species list to make space for a new one.
+  for (size_t i = 0; i < numSpecies; i++)
+    speciesList[i].clear();
+
+  // Assign the genomes to their species.
+  for (size_t i = 0; i < assignments.n_elem; i++)
+    speciesList[assignments[i]].push_back(genomeList[i]);
+}
+
+// Creates the next generation through reproduction.
+template <class TaskType,
+          class ActivationFunction,
+          class SelectionPolicy>
+void NEAT<TaskType, ActivationFunction, SelectionPolicy>::Reproduce()
+{
+  // The mean fitnesses of the species.
+  arma::vec meanFitnesses(numSpecies, arma::fill::zeros);
+  // The next generations sizes of the species.
+  arma::vec speciesSize(numSpecies, arma::fill::zeros);
+  // The number of elite membes in each species.
+  arma::vec numElite(numSpecies, arma::fill::zeros);
+
+  // Find the mean fitnesses.
+  for (size_t i = 0; i < numSpecies; i++)
+  {
+    if (speciesList[i].size() == 0)
+      continue;
+    for (size_t j = 0; j < speciesList[i].size(); j++)
+      meanFitnesses[i] += speciesList[i][j].Fitness();
+    meanFitnesses[i] /= speciesList[i].size();
+  }
+  double totalMeanFitness = arma::accu(meanFitnesses);
+
+  // Allot the sizes of the species.
+  for (size_t i = 0; i < numSpecies; i++)
+  {
+    speciesSize[i] = std::round(meanFitnesses[i] / totalMeanFitness * popSize);
+    // Handle special case where fitnesses are negative.
+    if (speciesSize[i] < 0)
+      speciesSize[i] = 0;
+  }
+
+  // If the total allotted size is less than the population, fill it.
+  int delta = (int)popSize - (int)arma::accu(speciesSize);
+  if (delta > 0)
+  {
+    size_t i = 0;
+    while (delta > 0)
+    {
+      speciesSize[i++]++;
+      delta--;
+      if(i >= numSpecies)
+        i = 0;
```

Comment:
I guess you could do this a bit faster by mathematically calculating the number of complete repetion and the remaining offset. Though your choice. Same apply for the else if part.